File: normalized Spectral Clustering - Project/Extra pre-implementations/sub project1/kmeans.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <limits.h>
#include <ctype.h>
#include <assert.h>


typedef struct {
    double *lst;
    int l;
} Point;

typedef struct {
    Point centroid;
    double * sumVector;
    int l;
} Cluster;

Point newPoint(char * c);
int countTheD(char * c);
int centroidsFresh(Cluster *c, int k);
void totalVec(Cluster c, Point centro);
double euclidian(Point p1 , Point p2);
Cluster* kmean(int maxIterations, Point* dataPointArr, Cluster* allClusters, int dataPointSize, int k);
void printCentroids(Cluster* finalClusters, int k, int d);
Cluster newClust(Point * pointArr,int x);
int eucNormNoSqrt(double *oldPoints, double *newPoints,Cluster *c,int cnt);
int isNumber(char number[]);

int main(int argc, char *argv[]) {
    char c[1000];
    int max_iter = 200;
    int num, rowLength,j,d_new, k,h,r,rr;
    Point * inputPoints;
    Cluster *allClusters, *finalClusters;
    rowLength = 20;
    inputPoints = (Point *) calloc(rowLength, sizeof(Point));
    num = 0;
    if ((argc<2)){
        printf("An Error Has Occurred");
        return 1;
    }
    if ((argc>3)){
        printf("An Error Has Occurred");
        return 1;
    }
    r = isNumber(argv[1]);
    if (r>0){
        printf("Invalid number of clusters!");
        return 1;
    }
    k = atoi(argv[1]);
    if (k<1){
        printf("Invalid number of clusters!");
        return 1;
    }


    if (argc == 3){
        rr = isNumber(argv[2]);
        if (rr>0){
            printf("Invalid number of iteration!");
            return 1;
        }
        max_iter = atoi(argv[2]);

        if ((max_iter >999)){
            printf("Invalid maximum iteration!");
            return 1;
        }
        if ((max_iter<2)){
            printf("Invalid maximum iteration!");
            return 1;
        }
    }

    if (inputPoints ==NULL){
        printf("An Error Has Occurred");
        return 1;
    } 


   while (fgets(c,1000, stdin) != NULL){
        if (num >= rowLength){
            inputPoints = (Point*) realloc(inputPoints , rowLength * 2 * sizeof(Point));
                if (inputPoints ==NULL){
                printf("An Error Has Occurred");
                return 1;
            }
            rowLength = rowLength * 2;
        }
        inputPoints[num] = newPoint(c);
        num = num + 1;
    }
    if (num<=k){
        printf("Invalid number of clusters!");
        return 1;  
    }


    allClusters = (Cluster *) malloc(k * sizeof(Cluster));
    if (allClusters ==NULL){
        printf("Invalid number of clusters!");
        return 1;
    } 
    for (j = 0 ; j < k; j++) {
        allClusters[j] = newClust(inputPoints, j);
    }
    d_new = inputPoints->l;
    finalClusters = (Cluster *) kmean(max_iter, inputPoints, allClusters, num, k);
    
    
    printCentroids(finalClusters, k, d_new);
    for ( h= 0; h < k; h++){
        free(allClusters[h].sumVector);
        free(allClusters[h].centroid.lst);
        allClusters[h].sumVector = NULL;
        allClusters[h].centroid.lst = NULL;
    }
    for ( h= 0; h < num; h++){
        free(inputPoints[h].lst);
        inputPoints[h].lst= NULL;
    }
    for ( h= 0; h < k; h++){
        free(finalClusters[h].sumVector);
        free(finalClusters[h].centroid.lst);
        finalClusters[h].sumVector = NULL;
        finalClusters[h].centroid.lst = NULL;
    }
    free(inputPoints);
    free(allClusters);
    allClusters = NULL;
    inputPoints = NULL;
    return 0;
}

int isNumber(char number[]){
    int i = 0;
    if (number[0] == '-'){
        i = 1;
    }
    for (; number[i] != 0; i++){
        if (!isdigit(number[i])){
            return 1;
        }
    }
    return 0;
}

Cluster newClust(Point * pointArr,int x){
    Cluster cluster;
    Point point;
    double *lst, *totalVec;
    int e;
    lst = (double *) malloc(pointArr[x].l * sizeof(double));
    for (e = 0 ; e < pointArr[x].l ; e++){
        lst[e] = pointArr[x].lst[e];
    }
    point.lst = lst;
    point.l = pointArr[x].l;
    totalVec = (double *) calloc(pointArr[x].l , sizeof(double));
    cluster.centroid = point;
    cluster.sumVector = totalVec;
    cluster.l = 0;
    return cluster;
}

Point newPoint(char * c){
    int size = countTheD(c) , j = 0;
    const char comma[2] = ",";
    double *float_lst;
    char* tok;
    Point dataPoint;
    float_lst = (double*) calloc(size, sizeof(double));
    tok = strtok(c , comma);
    while (tok != NULL) {
        float_lst[j] = atof(tok);
        tok = strtok(NULL, comma);
        j = j + 1;
    }
    dataPoint.lst = float_lst;
    dataPoint.l = size;
    return dataPoint;
}

int countTheD(char * c){
    int l = 0;
    char *t;
    t = c;
    while (*t != '\0'){
        if (*t == ','){
            l++;
        }
        t++;
    }
    return l+1;
}

void printCentroids(Cluster* finalClusters, int k, int d) {
    int a,b;
    for ( a = 0 ; a < k ; a++) {
        for ( b = 0 ; b < d  ; b++) {
            printf("%.4f", finalClusters[a].centroid.lst[b]);
            if (b != d-1){
                printf(",");
            }
        }
        printf("\n");
    }
}


Cluster * kmean(int max_iter, Point* inputArr, Cluster* allClusters, int d_new, int k) {
    int countModify = 3;
    int iterations = 0;
    int e,j,closest_Index;
    double min_dis, distance;
    while (iterations < max_iter && countModify>0) {
        iterations= iterations + 1;
        for ( e = 0; e < d_new; e++) {
            min_dis = INT_MAX; 
            closest_Index = -1;
            for ( j = 0 ; j < k; j++) {
                distance = euclidian(allClusters[j].centroid, inputArr[e]);
                if (distance < min_dis) {
                    min_dis = distance;
                    closest_Index = j;
                }
            }
            allClusters[closest_Index].l++;
            totalVec(allClusters[closest_Index], inputArr[e]);
        }
        countModify = centroidsFresh(allClusters, k);
    }
    return allClusters;
}

double euclidian(Point p1 , Point p2){
    double dis;
    int e;
    dis = 0;
    for ( e = 0; e < p1.l; e++){
        dis = dis + ((p1.lst[e] - p2.lst[e])*(p1.lst[e] - p2.lst[e]));
    }
    return dis;
}

void totalVec(Cluster c, Point centro){
    int e;
    for ( e = 0; e < centro.l; e++){
        c.sumVector[e] = c.sumVector[e] + centro.lst[e];
    }
}

int centroidsFresh(Cluster *c, int k){
    double *old_points;
    int cnt=0,e,j,m;
    for (e = 0; e < k; e++){
        old_points = (double*) calloc(c[e].centroid.l, sizeof(double));
        for ( j = 0; j < c[e].centroid.l; j++){
            old_points[j] = c[e].centroid.lst[j];
        }
        if (c[e].l != 0) {
            for (m = 0; m < c[e].centroid.l; m++) {
                c[e].centroid.lst[m] = (c[e].sumVector[m] / (double)c[e].l);
                c[e].sumVector[m] = 0;
            }
        }
        c[e].l = 0;
        cnt = eucNormNoSqrt(old_points, c[e].centroid.lst,c,cnt );
        free(old_points);
    }
    return cnt;  
}

int eucNormNoSqrt(double *oldPoints, double *newPoints,Cluster *c,int cnt){
    double normPow,epsilonPowerTwo;
    int e;
    normPow = 0;
    epsilonPowerTwo= 0.001*0.001;
    for ( e=0; e< c[e].centroid.l;e++){
        normPow = normPow +((newPoints[e])-(oldPoints[e]))*((newPoints[e])-(oldPoints[e]));
    }
    if (normPow >= epsilonPowerTwo){
        cnt = cnt+1;
    }
    return cnt;
}

File: normalized Spectral Clustering - Project/Extra pre-implementations/sub project1/kmeans.py
import sys

class Cluster :

    def __init__(self, datapoint) :
        self.centroid = datapoint
        self.dataPointsSum = [0] * len(datapoint)
        self.size = 0

    
    def addVectorToCluster(self, vector) :
        self.size += 1
        for i in range(len(vector)) :
            self.dataPointsSum[i] += vector[i]

    
    def euclideanDistance(self, vector) :
        squareSum = 0
        for i in range(len(vector)) :
            squareSum += (self.centroid[i] - vector[i]) ** 2
        return squareSum ** 0.5


    def getNewCentroid(self) :
        if (self.size == 0) :
            return self.centroid
        centroid =[]
        for i in range(len(self.dataPointsSum)) :
            centroid.append(self.dataPointsSum[i] / self.size)
        return centroid

    
    def updateCentroid(self, centroid) :
        self.centroid = centroid
        self.dataPointsSum = [0] * len(self.centroid)
        self.size = 0

    
def validInput(args) :
    if (len(args) != 2 and len(args) != 3) :
        print("An Error Has Occurred")
        return False
    if ((not args[0].isnumeric()) or int(args[0]) <= 1) :
        print("Invalid number of clusters!")
        return False
    if (len(args) == 3) :
        if ((not args[1].isnumeric()) or int(args[1]) <= 1 or int(args[1]) >= 1000) :
            print("Invalid maximum iteration!")
            return False
    return True


def readDataPoints(filePath) :
    try :
        dataPoints =[]
        file = open(filePath, "r")
        while (True) :
            line = file.readline()
            if (not line) :
                break
            line = line.rstrip('\n').split(",")
            for i in range(len(line)) :
                line[i] = float(line[i])
            dataPoints.append(line)
    except :
        sys.exit("An Error Has Occurred")
    file.close()
    return dataPoints


def initializeClusters(inputAsMat, k) :
    initialClusters =[]
    for i in range(k) :
        initialClusters.append(Cluster(inputAsMat[i]))
    return initialClusters


def updateClustersAndCheckConvergence(clusters) :
    eps = 10**(-3)
    converged = True
    for cluster in clusters :
        newCentroid = cluster.getNewCentroid()
        if ((converged) and (cluster.euclideanDistance(newCentroid) >= eps)) :
            converged = False
        cluster.updateCentroid(newCentroid)
    return converged


def kmeans(inputAsMat, k, iter) :
    clusters = initializeClusters(inputAsMat, k)
    iterations = 0
    while (iterations < iter) :
        iterations += 1
        for vector in inputAsMat :
            minDistance = float("inf")
            for i in range(len(clusters)) :
                currDst = clusters[i].euclideanDistance(vector)
                if (currDst < minDistance) :
                    minDistance = currDst
                    minClusterIndex = i
            clusters[minClusterIndex].addVectorToCluster(vector)
        if (updateClustersAndCheckConvergence(clusters)) :
            break
    return clusters


def printCentroids(clusters) :
    for cluster in clusters :
        centroidAsStr = ",".join(str('%.4f' % point) for point in cluster.centroid)
        print(centroidAsStr)


def main(args) :
    assert(validInput(args))
    k = int(args[0])
    if (len(args) == 2) :
        iter = 200
        filePath = args[1]
    else :
        iter = int(args[1])
        filePath = args[2]
    inputAsMat = readDataPoints(filePath)
    if (k >= len(inputAsMat)) :
        sys.exit("Invalid number of clusters!")
    clusters = kmeans(inputAsMat, k, iter)
    printCentroids(clusters)


if (__name__ == "__main__") :
    main(sys.argv[1:])